### ISSUE TITLE
LaTeX broken on current (2.12) spec rendering

### ISSUE THREAD

**Issue by JoergWMittag** [scala/scala.github.com#644]
Note: I'm not 100% sure this belongs here, but the main Scala issue tracker seems to be about *semantic* issues with the spec, not formatting, so I'm posting it here.

Currently, LaTeX rendering is broken on the current (2.12) spec. The issue seems to be that the LaTeX has been processed by the highlighter before it gets to the LaTeX script. I.e. [the HTML that gets sent to the browser looks like this](http://scala-lang.org/files/archive/spec/2.12/09-top-level-definitions.html):

```html
<div class="highlight"><pre><code class="language-scala" data-lang="scala"><span class="k">package</span> <span class="nn">$p_1$</span><span class="o">;</span>
<span class="n">$</span><span class="o">\</span><span class="n">ldots$</span>
<span class="k">package</span> <span class="nn">$p_n$</span><span class="o">;</span>
<span class="n">$</span><span class="o">\</span><span class="n">mathit</span><span class="o">{</span><span class="n">stats</span><span class="o">}</span><span class="n">$</span>
```

Note how the LaTeX expressions are broken up by separate `<span>` elements.

It [should look more like this](http://scala-lang.org/files/archive/spec/2.11/09-top-level-definitions.html) (the same fragment from the 2.11 spec, note that [the actual markdown source file for that fragment](https://github.com/scala/scala/blame/2.13.x/spec/09-top-level-definitions.md#L28-L33) is actually unchanged between the two):
```html
<div class="highlight"><pre><code class="language-scala" data-lang="scala">package $p_1$ { $\ldots$
  package $p_n$ {
    $\mathit{stats}$
  } $\ldots$
}
</code></pre></div>
```

As a result, the rendered code looks a bit like this:

>**`package`** *p<sub>1</sub>* `;`
>$\ldots$
>**`package`** *p<sub>n</sub>* `;`
>$\mathit{stats}$

instead of 

>**`package`** *p<sub>1</sub>* `;`
>…
>**`package`** *p<sub>n</sub>* `;`
>*stats*

Note that *some* LaTeX expressions are kept intact, e.g. `$p_1$` and `$p_n$` are not broken up into separate `<span>`s and thus rendered correctly; whereas `$\ldots$` and `$\mathit{stats}` are.

The only difference between 2.11 (working) and 2.12 (broken) I have been able to find at a quick glance is the Jekyll upgrade. [The actual markdown source for that fragment](https://github.com/scala/scala/blame/2.13.x/spec/09-top-level-definitions.md#L28-L33) hasn't been changed in years and looks like this:

    ```scala
    package $p_1$;
    $\ldots$
    package $p_n$;
    $\mathit{stats}$
    ```

[Discovered by [@hao luo on StackOverflow](http://stackoverflow.com/q/41280237/2988).]

**Comment by SethTisue** [comment]
is the problem reproducible when building the spec locally?

**Comment by JoergWMittag** [comment]
So, I just realized that 2.13 builds are up now on the website, and what do you know? They don't exhibit the bug: http://scala-lang.org/files/archive/spec/2.13/09-top-level-definitions.html

**Comment by JoergWMittag** [comment]
@SethTisue Yes, it is reproducible by checking out the default (2.12) branch and running 

```sh
bundle install
bundle exec jekyll serve -d build/spec/ -s spec/ -w --baseurl=""
```

as per the README.

**Comment by SethTisue** [comment]
> 2.13 builds are up now on the website, and what do you know? They don't exhibit the bug

that is downright weird. is there a volunteer out there who would like to look into this?

**Comment by SethTisue** [comment]
also reported at https://issues.scala-lang.org/browse/SI-10179

**Comment by janekdb** [comment]
If I can build the spec locally then I could take a look.

@JoergWMittag - Thanks for the build instructions above. Are these intended to be in the scala `README.md`?

**Comment by SethTisue** [comment]
@janekdb the `spec` directory has its own readme, https://github.com/scala/scala/blob/2.12.x/spec/README.md

**Comment by janekdb** [comment]
Just getting up to speed with Jekyll and Ruby gems! Aiming to match the versions here,

> https://pages.github.com/versions/

## Question 1

Noticed GH is using Jekyll 3.3.1 but `README.md` has a different view,

> We use Jekyll 2 and [Redcarpet](https://github.com/vmg/redcarpet) to generate the html. Essentially, this is what github pages use.

Just the README lagging?

## Question 2

I do not see `redcarpet` listed here: https://pages.github.com/versions/

Is that significant?

**Comment by SethTisue** [comment]
The situation w/r/t Markdown versions, Jekyll versions, and hosting, is a bit complicated.

scala-lang.org is hosted at EPFL and combines the contents of the scala/scala-lang repo with the contents of the spec directory in the scala/scala repo.

docs.scala-lang.org is hosted on GitHub Pages.

Whatever GitHub Pages uses is a pretty strong de facto standard in the Jekyll world, so we try to keep our three Jekyll configs in synch, but we're, uh, often not very successful at that. PRs reducing the drift between the three configs are highly encouraged.

**Comment by janekdb** [comment]
In pictures,

![scala-docs-repos](https://cloud.githubusercontent.com/assets/1123855/23064198/74b71bb6-f506-11e6-9f75-fc11fb0e1455.png)

And Jekyll 2 (not 3) is currently used at EPFL?

**Comment by janekdb** [comment]
Jekyll version updated in README.md: https://github.com/scala/scala/pull/5713

This issue should really be in `scala/scala`.

**Comment by janekdb** [comment]
TODO

 - [ ] Wait for 2.12.x to be merged to 2.13.x to get Jekyll update to 3.3.0 (https://github.com/scala/scala/pull/5715)

**Comment by janekdb** [comment]
These are examples of the two types of differences seen between the 2.11.x html and the 2.12.x html with `highlighter = false` instead of `null`

`div` is now `figure`

![spec-2 11-cf-2 12-highlighter-false-div-is-figure](https://cloud.githubusercontent.com/assets/1123855/23210006/8d7e3292-f8f3-11e6-93e3-99535fc6c34f.png)

Some character entities are now literals,

![spec-2 11-cf-2 12-highlighter-false-entities-gone](https://cloud.githubusercontent.com/assets/1123855/23210021/9c0affc0-f8f3-11e6-99e8-7727c08517c8.png)

**Comment by janekdb** [comment]
https://github.com/scala/scala/pull/5731 with the `highlighter = false` change.

**Comment by janekdb** [comment]
Side by side visual comparison shows only a layout change on the figures.

https://www.scala-lang.org/files/archive/spec/2.11/09-top-level-definitions.html
https://www.scala-lang.org/files/archive/spec/2.12/09-top-level-definitions.html

In the 2.12 version the grey figure box is narrower and has more whitespace around it. The 2.11 layout is better so this is a retrograde step but an acceptable loss given the main problem is now fixed.

@JoergWMittag: Suggest to `close`.

**Comment by SethTisue** [comment]
Janek, you deserve a medal for this. Seriously, thanks for tracking this one down.

**Comment by JoergWMittag** [comment]
I hate to rub salt into this wound, but I just discovered that now the 2.12 version works, but the 2.13 version is broken, i.e. the opposite of what we started out with:

* 2.12: http://scala-lang.org/files/archive/spec/2.12/09-top-level-definitions.html
* 2.13: http://scala-lang.org/files/archive/spec/2.13/09-top-level-definitions.html

**Comment by lrytz** [comment]
I didn't follow things in detail here, but this might be because the PR https://github.com/scala/scala/pull/5731 was not yet merged into 2.13.x, it will be soon though: https://github.com/scala/scala/pull/5759
